File: examples/raphtory-example-twitter/src/main/scala/com/raphtory/examples/twitter/higgsdataset/analysis/MemberRank.scala
package com.raphtory.examples.twitter.higgsdataset.analysis

import com.raphtory.api.algorithm.Generic
import com.raphtory.api.graphview.GraphPerspective
import com.raphtory.api.table.Row
import com.raphtory.api.table.Table
import com.raphtory.api.visitor.Edge

/**
  * Description
  * This algorithm takes the Page Rank Score from the vertices neighbour's and multiplies
  * this by the vertices ranking from the original raw dataset (the vertex in degree). If the vertex gets a high MemberRank score,
  * this means that other people that have high Page Rank values have rated them highly. This should dampen
  * the effects of bots further.
  */

class MemberRank() extends Generic {

  case class Score(negativeScore: Double = 0.0, positiveScore: Double = 0.0) {

    def +(that: Score): Score = {
      val negative = this.negativeScore + that.negativeScore
      val positive = this.positiveScore + that.positiveScore

      Score(negative, positive)
    }
  }

  override def apply(graph: GraphPerspective): graph.Graph =
    graph
      .step { vertex =>
        //For this vertex get our Page Rank value
        val prValue = vertex.getState[Double]("prlabel")
        //Message all of our outgoing neighbours with our ID + Page Rank value
        vertex.messageOutNeighbours((vertex.ID(), prValue))
      }
      .step { vertex =>
        //Get the Vertex ID and Page Rank (Long,Double) messages from our neighbours
        val queue: Seq[(vertex.IDType, Double)] = vertex.messageQueue[(vertex.IDType, Double)]

        //Get the vertex in degree
        val inDegree = vertex.getInNeighbours().size.toDouble

        //makes list of all the raw scores from data
        val rawScoreList: List[Score] = vertex.getInEdges().map {
          case edge: Edge =>
            if (inDegree < 0) Score(negativeScore = inDegree)
            else Score(positiveScore = inDegree)
          case _          => Score()
        }

        //Take output of val rawScoreList, sum all the positive scores and position on left of tuple, sum all negative scores and position on right of tuple. Does this for each vertex.
        val totalRaw: Option[Score] = rawScoreList
          .reduceLeftOption((scoreLeft, scoreRight) => scoreLeft + scoreRight)

        //Extract negative and positive scores from totalRaw tuple.
        val negativeRawScore = totalRaw.fold(0.0)(_.negativeScore)
        val positiveRawScore = totalRaw.fold(0.0)(_.positiveScore)

        val newScore: Seq[Score] = queue.map {
          case (neighbourID, prScore) =>
            vertex.getInEdge(neighbourID) match {
              case Some(edge) =>
                //Multiplies page rank score of neighbour (how popular it is) to its own score
                if (inDegree < 0) Score(negativeScore = inDegree * prScore)
                else Score(positiveScore = inDegree * prScore)

              case None       => Score()
            }
        }

        //Take output of val newScore, sum all the positive scores and position on left of tuple, sum all negative scores and position on right of tuple. Does this for each vertex.
        val totalScore: Option[Score] = newScore
          .reduceOption((scoreLeft, scoreRight) => scoreLeft + scoreRight)

        //Extract negative and positive scores from totalScore tuple.
        val negativeNewScore = totalScore.fold(0.0)(_.negativeScore)
        val positiveNewScore = totalScore.fold(0.0)(_.positiveScore)

        //set's vertex state with these scores
        vertex.setState("negativeRawScore", negativeRawScore)
        vertex.setState("negativeNewScore", negativeNewScore)
        vertex.setState("positiveRawScore", positiveRawScore)
        vertex.setState("positiveNewScore", positiveNewScore)
      }

  /**
    * Reports the following:
    *   1) Vertex ID
    *   2) Page Rank Score
    *   3) Original Negative Score
    *   4) Original Positive Score
    *   5) New Negative Score
    *   6) New Positive Score
    */
  override def tabularise(graph: GraphPerspective): Table =
    graph.select { vertex =>
      Row(
              vertex.getPropertyOrElse("name", vertex.ID()),
              //a vertices page rank score
              vertex.getStateOrElse("prlabel", -1),
              //gets vertex states and tabularises
              vertex.getStateOrElse("negativeRawScore", 0.0),
              vertex.getStateOrElse("positiveRawScore", 0.0),
              vertex.getStateOrElse("negativeNewScore", 0.0),
              vertex.getStateOrElse("positiveNewScore", 0.0)
      )
    }
}

object MemberRank {
  def apply() = new MemberRank()
}

File: examples/twitter/src/main/python/memberrank.py
from pyraphtory.algorithm import PyAlgorithm
from pyraphtory.graph import TemporalGraph, Row, Table
from pyraphtory.vertex import Vertex
from dataclasses import dataclass
import pyraphtory.scala.collection
from functools import reduce

@dataclass(frozen=True)
class Score(negativescore = 0.0, positivescore = 0.0):
    def plus(self, that):
        negative = self.negativescore + that.negativescore
        positive = self.positivescore + that.positivescore
    
        Score(negative,positive)


class MemberRank(PyAlgorithm):
    def __init__(self):
        pass

    def __call__(self, graph: TemporalGraph) -> TemporalGraph:
        def init(v: Vertex):
            prvalue = v.get_state('prlabel')
            v.message_out_neighbours(v.id, prvalue)
            

        def step(v: Vertex):
            queue = v.message_queue
            indegree = v.in_degree.float()
            rawscorelist = v.in_edges()
            match map(rawscorelist):
                case graph.edge:
                    if(indegree < 0.0): 
                        Score(negativescore = indegree)
                    else:
                        Score(positivescore = indegree)
                case _:
                    Score()

            totalraw = reduce(lambda x,y:(x[0] + y[0], x[1] + y[1]), rawscorelist)

            negativerawscore = reduce(lambda x,y: x, totalraw)
            positiverawscore = reduce(lambda x,y: y, totalraw)

            match map(queue):
                case (neighbourid, prscore):
                    edge = v.get_in_edge(neighbourid)
                    if (indegree < 0): 
                       newscore = Score(negativescore = indegree * prscore)
                    else: 
                       newscore = Score(positivescore = indegree * prscore)   
                case _:
                    newscore = Score()

            totalscore = reduce(lambda x,y:(x[0] + y[0], x[1] + y[1]), newscore)
            negativenewscore = reduce(lambda x,y: x, totalscore)
            positivenewscore = reduce(lambda x,y: y, totalscore)

            v.set_state('negativerawscore', negativerawscore)
            v.set_state('negativenewscore', negativenewscore)
            v.set_state('positiverawscore', positiverawscore)
            v.set_state('positivenewscore', positivenewscore)

        def tabularise(self, graph: TemporalGraph):
            def get_scores(v: Vertex):
                rows_found = Row(
                                v.get_property_or_else('name', v.id),
                                v.get_state_or_else('prlabel', -1),
                                v.get_state_or_else("negativeRawScore", 0.0),
                                v.get_state_or_else("positiveRawScore", 0.0),
                                v.get_state_or_else("negativeNewScore", 0.0),
                                v.get_state_or_else("positiveNewScore", 0.0)
                            )
                return rows_found
            return graph.select(lambda v: get_scores(v))
        